### ISSUE TITLE
Negative .632+ Bootstrap MSEs

### ISSUE THREAD

**Issue by javdg** [mlr-org/mlr#512]
As requested by @berndbischl after discussing the issue via email, the following is a (minimal) example which should produce a negative .632+ bootstrap mse (see cv iter 10) when tuning a regr.glmnet ridge regression model (using mlr's current devel version). The simulated data is approximating the properties of a considerably larger (both in terms of p and n, I tried to reduce complexity), real life data set.

``` R
library(mlr)
library(glmnet)

n <- 100
cols <- as.list(letters[1:6])
d <- matrix(nrow = n, ncol = length(cols))
colnames(d) <- cols

set.seed(1)
d[, 1] <- round(rnorm(n, 80, 10))
d[, 2] <- rbinom(n, 1, .7)
d[, 3] <- rpois(n, .7) + 15
d[, 4] <- round(rnorm(n, 15, 2))
d[, 5] <- round(rnorm(n, 12, 3))
d[, 6] <- round(rnorm(n, 13, 3))

d <- as.data.frame(d)

# task
tsk <- makeRegrTask(id = "task_test", data = d, target = "a")

# inner loop
inner <- makeResampleDesc("Bootstrap", iters = 10, predict = "both")
mse_b632plus <- setAggregation(mse, b632plus)

ctrl <- makeTuneControlGrid()

# learner
ps_ridge <- makeParamSet(
    makeDiscreteParam("lambda", 10^seq(10, -2, length.out = 10)),
    makeDiscreteParam("alpha", 0)
)

lrn_ridge <- makeLearner("regr.glmnet", "ridge")
lrn_ridge <- makeTuneWrapper(lrn_ridge, resampling = inner, measures = mse_b632plus, par.set = ps_ridge, control = ctrl)

# outer loop
outer <- makeResampleDesc("CV", iters = 10)

bmr <- benchmark(lrn_ridge, tsk, outer, measures = mse)

getBMRTuneResults(bmr, as.df = TRUE)

```

```
> getBMRTuneResults(bmr, as.df = TRUE)
     task.id  learner.id iter lambda alpha mse.b632plus
1  task_test ridge.tuned    1  1e+02     0  57.08658900
2  task_test ridge.tuned    2  1e+02     0  54.07339975
3  task_test ridge.tuned    3  1e+02     0  42.50338371
4  task_test ridge.tuned    4  1e+02     0  51.16843158
5  task_test ridge.tuned    5  1e+02     0   0.05424766
6  task_test ridge.tuned    6  1e-02     0  51.33880670
7  task_test ridge.tuned    7  1e+02     0  69.74140824
8  task_test ridge.tuned    8  1e+02     0  35.08206086
9  task_test ridge.tuned    9  1e-02     0   4.84825097
10 task_test ridge.tuned   10  1e-02     0  -6.44854242
```

I was/am wondering if this is due to some error in my code, incorrect model assumptions/a misspecification, unfeasible data, or another, possibly more fundamental computational problem?
I'm happy to provide additional information if needed.

Cheers,
Johannes

**Comment by larskotthoff** [comment]
Hmm, this actually looks reasonable to me. B632+ estimates the relative overfitting of the model with respect to the no-information case where the features do not inform the prediction at all. To do this, is subtracts the training error from the test error. A negative value simply means that the training error was larger than the test error in this case (i.e. you have a "negative overfitting").

The large values in the other folds are somewhat worrying, but the negative value is fine.

**Comment by javdg** [comment]
Thanks Lars, though I have to say I'm not thoroughly convinced yet.

I had the (possibly amateurish) impression, that .632+ bootstrap estimates are based on a weighted sum of training and test set error (in the same line of thought as the .632 bootstrap) , which I would not expect to be negative.
Taking another look at Efron & Tibshirani (1997, http://cindy.informatik.uni-bremen.de/cosy/teaching/CM_2011/Eval3/efron_97.pdf) I understand the .632+ bootstrap estimate to be presented in formula (29), "a less biased version of (24) [the .632 bootstrap estimate], where the weights on err and ^Err1 depend on ^R", "a quantity that ranges from 0 [...] to 1 [...]" (cf. formula (28)). This leads to the "weight w [in formula (29)] rang[ing] from .632 if ^R = 0 to 1 if ^R = 1" and the conclusion is, that the .632+ bootstrap estimate "ranges from ^Err.632 [the .632 bootstrap estimate] to ^Err1 [the, as I understand it, test set error]". None of this would lead me to expect negative values.
However, Efron & Tibshirani seem to acknowledge 'edge cases' (my words, not theirs, so this might be rather common) "in which ^R can fall outside of [0, 1]" and I suspect this could produce negative outcomes. They present a correction (^R') in formula (31) which is used for their following calculations - see formula (32). Again, to my understanding, this is preventing the occurrence of negative outcomes.

Could this adjustment be missing in mlr's implementation?

Thanks for your support!

Disclaimer: I'm not a *true* statistician, so I might have gotten everything of the above terribly wrong. :)

**Comment by berndbischl** [comment]
Hi, I will need some time to review this properly, so a proper answer might arrive next. 
But I vaguely seem to remember what you are talking about. If we can adapt the formula here easily, maybe via a user option, we might do that. I need to review your example, our code and the Efron paper simultaneously for that.

**Comment by larskotthoff** [comment]
@javdg You're correct in your analysis. What's implemented in mlr is basically (29) from that paper, with the overfitting rate computed according to (28). It's easy to see from this equation that if the training error is larger than the testing error, the whole thing becomes negative, which is what's happening in your case.

mlr doesn't implement the correction they propose. I personally don't think it should either. From a theoretical point of view, it certainly makes sense to assume that the worst error is the one you get in the no-information case (which is what their correction does), but in practice it can be much worse if you have a terrible model. And in practice you probably want to know whether you have such a terrible model.

In the end it depends how you interpret and use the value. If the relative overfitting is 0, you may conclude that you have a model that generalises well when the model is actually essentially completely random and makes worse predictions that random guessing. By applying the correction proposed in the paper, you're "masking" this behaviour (because the relative overfitting can't be less than 0).

That said, this weighted error doesn't tell you the whole story anyway and you always have to look at absolute performance values etc as well, so maybe adding the correction would be fine (although I think that it would be misleading). It's certainly easy enough to code.

**Comment by larskotthoff** [comment]
@berndbischl Do we want to include this in 2.9?

**Comment by javdg** [comment]
As per our conversation from the weekend, this is to ping @berndbischl ...

**Comment by zmjones** [comment]
@berndbischl why is this labelled severe? are we waiting on someone to implement the correction? do you disagree with @larskotthoff regarding his above comment?

**Comment by larskotthoff** [comment]
@berndbischl ping
